### PR TITLE
Adding support for custom metric logging via LogMetric

### LINKED ISSUE #543: ILogger missing LogMetric extension method
In Functions v2 (in proc) we had [an extension method defined in `Microsoft.Azure.Webjobs.Host` that allowed logging of AppInsights `metric` types](https://github.com/Azure/azure-webjobs-sdk/blob/206e194e03d82a101a6b5dffb7b4c3904730801d/src/Microsoft.Azure.WebJobs.Host/Loggers/Logger/LoggerExtensions.cs) but this doesn't exist in the new `Worker.*` namespaces. We should bring it forward.

### PR DESCRIPTION
**Pull request by brandonh-msft**
Adds a new `LogMetric()` extension method which sends a new `RpcMetric` type through the gRPC stream to the host. Contains all the information needed to log the metric as a custom metric on the host side.

Changes should be merged only after https://github.com/Azure/azure-functions-host/pull/7593

Fixes #543

### CODE REVIEW

**Comment by brandonh-msft** [comment]
This at least provides the extension, but we still need something [like this](https://github.com/Azure/azure-webjobs-sdk/commit/e6babfd91fc03bbe9e534a90c61fab290b85a9ab) to actually log things _as metrics_ in AppInsights.  I couldn't find where in the codebase we're actually sending stuff to `TelemetryClient`

**Comment by brandonh-msft** [comment]
Changes should be merged after https://github.com/Azure/azure-functions-host/pull/7593

**Comment by brandonh-msft** [comment]
😊 worth a shot

File: src/DotNetWorker.Core/Context/FunctionContextLoggerExtensions.cs
Comment (liliankasem) [inline src/DotNetWorker.Core/Context/FunctionContextLoggerExtensions.cs:34]: Just wanna understand why we're doing this:

`CreateFunctionUserCategory` -> The category for logs written from within user functions. Are all the logs here user logs? Curious why we need to wrap the categoryName this way

**Review by liliankasem — COMMENTED** [review]
@brandonh-msft some questions for you!

File: src/DotNetWorker.Core/Logging/LogConstants.cs
Comment (brandonh-msft) [reply src/DotNetWorker.Core/Logging/LogConstants.cs:12]: This is a bring-forward of work already present in the other Functions Host code; I've a feeling the rest of these are going to start being needed in the future but if you want to trim just for this PR it's your call :)

File: src/DotNetWorker.Core/Context/FunctionContextLoggerExtensions.cs
Comment (brandonh-msft) [reply src/DotNetWorker.Core/Context/FunctionContextLoggerExtensions.cs:34]: This was a difference I spotted in the previous .NET Worker code and so made it match our previous logging behavior.

Comment (brandonh-msft) [reply src/DotNetWorker.Core/Context/FunctionContextLoggerExtensions.cs:34]: Also yes. When you call `.GetLogger()` it's because you're the user and you're going to do some logging.

File: samples/Configuration/Configuration.csproj
Comment (liliankasem) [inline samples/Configuration/Configuration.csproj:12]: @brettsam Should we bump this to 1.3.0?

Comment (brettsam) [reply samples/Configuration/Configuration.csproj:12]: Yeah, might as well get it up on the latest whenever we see it falling behind.

File: src/DotNetWorker.Core/Context/FunctionContextLoggerExtensions.cs
Comment (brettsam) [inline src/DotNetWorker.Core/Context/FunctionContextLoggerExtensions.cs:34]: This doesn't seem like something we'd want to do... it'd change the category that someone explicitly passed into us.

File: src/DotNetWorker.Grpc/GrpcFunctionsHostLogger.cs
Comment (brettsam) [inline src/DotNetWorker.Grpc/GrpcFunctionsHostLogger.cs:22]: Instead of taking the serializer here -- have it injected via the constructor. That allows a customer to change it (see the Configuration sample in the repo for how they'd do that). They can change from Newtonsoft to System.Text.Json, for example.

Comment (liliankasem) [reply src/DotNetWorker.Grpc/GrpcFunctionsHostLogger.cs]: @brettsam Done! I followed your pattern of doing it through a set function instead of the constructor in `GrpcFunctionsHostLoggerProvider`, the same way `scopeProvider` is currently being set. Is that okay or should I change that?

Also just curious why `_scopeProvider` wasn't being set in the constructor the same way `_channelWriter` is when it is also a required parameter for `GrpcFunctionsHostLogger` and cannot be null.

Comment (liliankasem) [reply src/DotNetWorker.Grpc/GrpcFunctionsHostLogger.cs]: answered offline

**Comment by fabiocav** [comment]
Flagging this as blocked for clarity

**Comment by brandonh-msft** [comment]
Need anything from me?

**Comment by liliankasem** [comment]
> Need anything from me?

I don't think so, just waiting for a new host release

File: src/DotNetWorker.Core/Logging/ILoggerExtensions.cs
Comment (kshyju) [inline src/DotNetWorker.Core/Logging/ILoggerExtensions.cs:16]: Should we explicitly specify that we will be logging as Information level message? Either in the method comment/documentation or the method name?

File: src/DotNetWorker.Grpc/GrpcFunctionsHostLogger.cs
Comment (kshyju) [inline src/DotNetWorker.Grpc/GrpcFunctionsHostLogger.cs:91]: nit: invert `if` to reduce nesting.

File: src/DotNetWorker.Grpc/GrpcFunctionsHostLoggerProvider.cs
Comment (kshyju) [inline src/DotNetWorker.Grpc/GrpcFunctionsHostLoggerProvider.cs:29]: [In other places, we throw](https://github.com/Azure/azure-functions-dotnet-worker/blob/main/src/DotNetWorker.Core/Converters/JsonPocoConverter.cs#L26) when workerOptions.Serializer is `null`.

File: test/DotNetWorkerTests/Diagnostics/GrpcHostLoggerTests.cs
Comment (brandonh-msft) [reply test/DotNetWorkerTests/Diagnostics/GrpcHostLoggerTests.cs:109]: My best guess 6mo later is because the Metric log event's ID is `1` now (see `LogConstants`) and so system log id values needed to be different.

File: src/DotNetWorker.Grpc/GrpcFunctionsHostLogger.cs
Comment (brandonh-msft) [reply src/DotNetWorker.Grpc/GrpcFunctionsHostLogger.cs:108]: this code is simply mimicking what's already in `Log` but for metrics; if we could reuse the code in `Log` somehow that's cool

File: src/DotNetWorker.Core/Logging/LogCategories.cs
Comment (brandonh-msft) [reply src/DotNetWorker.Core/Logging/LogCategories.cs:4]: From what I recall this was used by something else in the pipeline and served to create system vs user log categories.

Comment (brandonh-msft) [reply src/DotNetWorker.Core/Logging/LogCategories.cs:4]: found it
https://github.com/Azure/azure-functions-host/blob/852eed9ceef6ef56b431428a8eb31f1bd9c97f3b/src/WebJobs.Script.WebHost/Diagnostics/SystemLoggerProvider.cs#L52

File: src/DotNetWorker.Grpc/GrpcFunctionsHostLoggerProvider.cs
Comment (brandonh-msft) [reply src/DotNetWorker.Grpc/GrpcFunctionsHostLoggerProvider.cs:23]: who do we envision "you" is addressing now that this PR is the better part of 1yr old? :)

Comment (liliankasem) [reply src/DotNetWorker.Grpc/GrpcFunctionsHostLoggerProvider.cs:23]: I'll be addressing PR changes, just needed you to answer some of the why questions :)

File: src/DotNetWorker.Grpc/GrpcFunctionsHostLogger.cs
Comment (brettsam) [inline src/DotNetWorker.Grpc/GrpcFunctionsHostLogger.cs:48]: Can we check the event *name* here instead of the *id*? Then we can remove that `-1` change in the test. Otherwise, I worry this is a breaking change and that anyone currently using an event with id of `1` will now go down this path.

**Review by brettsam — APPROVED** [review]

**Pull request merged**
